src/controllers/hid/hidcontroller.cpp

```diff
@@ -136,6 +136,69 @@ int HidController::close() {
+    Trace process("HidController processInputReport");
+    unsigned char* pPreviousBuffer = m_pPollData[(m_PollingBufferIndex + 1) % kNumBuffers];
+    unsigned char* pCurrentBuffer = m_pPollData[m_PollingBufferIndex];
+    // Some controllers such as the Gemini GMX continuously send input reports even if it
+    // is identical to the previous input report. If this loop processed all those redundant
+    // input reports, it would be a big performance problem to run JS code for every input report
+    // plus it would be unnecessary.
+    // This assumes that the redundant input reports all use the same report ID. In practice we
+    // have not encountered any controllers that send redundant input reports with different report
+    // IDs. If any such devices exist, this may be changed to use a separate buffer to store
+    // the last input report for each report ID.
+    if (bytesRead == m_LastPollSize &&
+            memcmp(pCurrentBuffer, pPreviousBuffer, bytesRead) == 0) {
+        return;
+    }
+    // Cycle between buffers so the memcmp below does not require deep copying to another buffer.
+    m_PollingBufferIndex = (m_PollingBufferIndex + 1) % kNumBuffers;
+    m_LastPollSize = bytesRead;
+    auto incomingData = QByteArray::fromRawData(
+            reinterpret_cast<char*>(pCurrentBuffer), bytesRead);
+    receive(incomingData, mixxx::Time::elapsed());
+}
+
+QList<int> HidController::getInputReport(unsigned int reportID) {
+    Trace hidRead("HidController getInputReport");
+    int bytesRead;
+
+    m_pPollData[m_PollingBufferIndex][0] = reportID;
+    bytesRead = hid_get_input_report(m_pHidDevice, m_pPollData[m_PollingBufferIndex], kBufferSize);
+
+    controllerDebug(bytesRead
+            << "bytes received by hid_get_input_report" << getName()
+            << "serial #" << m_deviceInfo.serialNumber()
+            << "(including one byte for the report ID:"
+            << QString::number(static_cast<quint8>(reportID), 16)
+                       .toUpper()
+                       .rightJustified(2, QChar('0'))
+            << ")")
+
+            if (bytesRead <= kReportIdSize) {
+        // -1 is the only error value according to hidapi documentation.
+        // Otherwise minimum possible value is 1, because 1 byte is for the reportID,
+        // the smallest report with data is therefore 2 bytes.
+        DEBUG_ASSERT(bytesRead <= kReportIdSize);
+        return QList<int>();
+    }
+
+    // Convert array of bytes read in a JavaScript compatible return type
+    // For compatibilty with the array provided by HidController::poll the reportID is contained as prefix
+    QList<int> dataList;
+    dataList.reserve(bytesRead);
+    for (int i = 0; i < bytesRead; i++) {
+        dataList.append(m_pPollData[m_PollingBufferIndex][i]);
+    }
+
+    // Execute callback function in JavaScript mapping
+    // and print to stdout in case of --controllerDebug
+    HidController::processInputReport(bytesRead);
```

Comment:
```suggestion
    processInputReport(bytesRead);
```
The class name is redundant inside a member function.

Reply:
Done